[CUDA][NCCL] Use nccl-cu13 instead of nccl-cu11 when building from source

nWEIdia opened the pull request:
with NCCL submodule.  BTW Remove all nccl-cu11 reference.

Fixes an issue noticed by @kwen2501 (thanks!) https://github.com/pytorch/pytorch/pull/168091#issuecomment-3629625108 

NB: one or more commits were generated by AI Agents. See commit message for details. 

cc @kwen2501 @fduwjj @atalman @malfet @tinglvv @eqy @ptrblck

nWEIdia committed:
[CUDA][NCCL] Use nccl-cu13 instead of nccl-cu11 when building from
source with NCCL submodule.  BTW Remove all nccl-cu11 reference.

kwen2501 commented on tools/optional_submodules.py:
```diff
@@ -29,8 +29,8 @@ def _checkout_by_tag(repo: str, tag: str) -> None:
-    if os.getenv("DESIRED_CUDA", os.getenv("CUDA_VERSION", "")).startswith("11"):
-        nccl_file = "nccl-cu11.txt"
+    if os.getenv("DESIRED_CUDA", os.getenv("CUDA_VERSION", "")).startswith("13"):
+        nccl_file = "nccl-cu13.txt"
```
I tried something like this but it doesn't seem to work -- there is no `DESIRED_CUDA` or `CUDA_VERSION` in the environment with the default build command. User has to add those.

nWEIdia replied:
Thank you! I'm thinking we could make cu13 the default, but it is not clear from this https://github.com/pytorch/pytorch/issues/165111 whether cu13 is considered stable. 
Given today we had to revert cu13 CI, it is probably ok to say cu12 is still the stable and requiring users to provide CUDA_VERSION for source building cu13 is acceptable.

kwen2501 commented:
> Can you please also upgrade the NCCL version for CUDA 12?

Checking CI in this PR: https://github.com/pytorch/pytorch/pull/169927
cc @nWEIdia

nWEIdia committed:
Detect CUDA version before checking out NCCL when doing source building.
Thanks Ke for this suggestion.
(The Real) Commit Author: AI Agents
I still perform the task of prompting for now.
